dataplug-facebook/app/com/hubofallthings/dataplugFacebook/apiInterfaces/authProviders/FacebookProvider.scala

```diff
@@ -0,0 +1,145 @@
+/*
+ * Copyright (C) 2020 Dataswift Ltd - All Rights Reserved
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ * Written by Marios Tsekis <marios.tsekis@dataswift.io> 5, 2020
+ */
+
+package com.hubofallthings.dataplugFacebook.apiInterfaces.authProviders
+
+import com.hubofallthings.dataplug.apiInterfaces.authProviders.HatOAuth2Provider
+import com.mohiva.play.silhouette.api.LoginInfo
+import com.mohiva.play.silhouette.api.util.HTTPLayer
+import com.mohiva.play.silhouette.impl.exceptions.ProfileRetrievalException
+import com.mohiva.play.silhouette.impl.providers._
+import com.mohiva.play.silhouette.impl.providers.oauth2.FacebookProvider._
+import play.api.libs.json.{ JsObject, JsValue }
+
+import scala.concurrent.Future
+
+/**
+ * Base Facebook OAuth2 Provider.
+ *
+ * @see https://developers.facebook.com/tools/explorer
+ * @see https://developers.facebook.com/docs/graph-api/reference/user
+ * @see https://developers.facebook.com/docs/facebook-login/access-tokens
+ */
+trait BaseFacebookProvider extends HatOAuth2Provider {
+
+  /**
+   * The content type to parse a profile from.
+   */
+  override type Content = JsValue
+
+  /**
+   * The provider ID.
+   */
+  override val id = ID
+
+  /**
+   * Defines the URLs that are needed to retrieve the profile data.
+   */
+  override protected val urls = Map("api" -> settings.apiURL.getOrElse(API))
+
+  /**
+   * Builds the social profile.
+   *
+   * @param authInfo The auth info received from the provider.
+   * @return On success the build social profile, otherwise a failure.
+   */
+  override protected def buildProfile(authInfo: OAuth2Info): Future[Profile] = {
+    httpLayer.url(urls("api").format(authInfo.accessToken)).get().flatMap { response =>
+      val json = response.json
+      (json \ "error").asOpt[JsObject] match {
+        case Some(error) =>
+          val errorMsg = (error \ "message").as[String]
+          val errorType = (error \ "type").as[String]
+          val errorCode = (error \ "code").as[Int]
+
+          throw new ProfileRetrievalException(SpecifiedProfileError.format(id, errorMsg, errorType, errorCode))
+        case _ => profileParser.parse(json, authInfo)
+      }
+    }
+  }
+}
+
+/**
+ * The profile parser for the common social profile.
+ */
+class FacebookProfileParser extends SocialProfileParser[JsValue, CommonSocialProfile, OAuth2Info] {
+
+  /**
+   * Parses the social profile.
+   *
+   * @param json     The content returned from the provider.
+   * @param authInfo The auth info to query the provider again for additional data.
+   * @return The social profile from given result.
+   */
+  override def parse(json: JsValue, authInfo: OAuth2Info) = Future.successful {
```

Comment:
A question: Are we parsing JSON like this, rather than using an extractor for historical reasons?

We can push all of the parsing into one place if we move over to extractors.

Reply:
To be honest I don't know what an extractor is or how it makes a difference. I would love to hear about it if you think we can benefit. 

If you are referring to the the `parse` function, that is the function Silhouette calls to read the profile from the profile URL of the given plug. This is called after a successful auth and it used to form the `CommonSocialProfile ` . We are extensively using the `loginInfo` since this is what every endpoint that is defined as a  `SecuredRequest` one, has to have. Usually with the session cookie that we store the token and then via the token we can extract and create the Information needed, eg HAT address, token with the 3rd party service, endpoint requests etc 

If you are referring to the way we read the json itself, we are using the data \ "path" in all of the project. You can read more about it here -> https://www.playframework.com/documentation/2.8.x/ScalaJson

Reply:
Fair, I prefer relying on types and then producing an extractor to do the work.  

It's not super important, but it's a bit cleaner.

The code here is known to work, and that's good enough for me. :)

Reply:
Are you referring to something like [this](https://github.com/Hub-of-all-Things/DataPlugs/blob/master/dataplug-gcalendar/app/com/hubofallthings/dataplugCalendar/models/GoogleCalendarEvent.scala)?

Usage can be found [here](https://github.com/Hub-of-all-Things/DataPlugs/blob/0c19fd43d6c488037d819e5fb2c6808d731efa6d/dataplug-gcalendar/app/com/hubofallthings/dataplugCalendar/apiInterfaces/GoogleCalendarEventsInterface.scala#L110l)

Reply:
Yes! Takes a bit more time to do it, but I think it's worth it.

Reply:
Totally. We do use the extractor pattern/method/way in all of our services. The difference with this one is that we didn't have a FacebookProvider and I took the original one, modified it to use the v5.0 of the API and left the rest unchanged. We can definitely write it cleaner and it's not something time consuming. I will add it to the list of things to change 👍